qiita_db/handlers/oauth2.py

```diff
@@ -0,0 +1,306 @@
+            _oauth_error(self, 'Oauth2 error: token has timed out',
+                         'invalid_grant')
+            return
+        # Check daily rate limit for key if password style key
+        if db_token['grant_type'] == 'password':
+            limit_key = '%s_%s_daily_limit' % (db_token['client_id'],
+                                               db_token['user'])
+            limiter = r_client.get(limit_key)
+            if limiter is None:
+                # Set limit to 5,000 requests per day
+                r_client.setex(limit_key, 5000, 86400)
+            else:
+                r_client.decr(limit_key)
+                if int(r_client.get(limit_key)) <= 0:
+                    _oauth_error(
+                        self, 'Oauth2 error: daily request limit reached',
+                        'invalid_grant')
+                    return
+
+        return f(self, *args, **kwargs)
+    return wrapper
+
+
+class OauthBaseHandler(RequestHandler):
+    def write_error(self, status_code, **kwargs):
+        """Overriding the default write error in tornado RequestHandler
+
+        Instead of writing all errors to stderr, this writes them to the logger
+        tables.
+
+        Parameters
+        ----------
+        status_code : int
+            HTML status code of the error
+        **kwargs : dict
+            Other parameters describing the error
+
+        Notes
+        -----
+        This function is automatically called by the tornado package on errors,
+        and should never be called directly.
+        """
+        if status_code in {403, 404, 405}:
+            # We don't need to log these failues in the logging table
+            return
+        # log the error
+        exc_info = kwargs['exc_info']
+        trace_info = ''.join(['%s\n' % line for line in
+                             format_exception(*exc_info)])
+        req_dict = self.request.__dict__
+        # must trim body to 1024 chars to prevent huge error messages
+        req_dict['body'] = req_dict.get('body', '')[:1024]
+        request_info = ''.join(['<strong>%s</strong>: %s\n' %
+                               (k, req_dict[k]) for k in
+                                req_dict.keys() if k != 'files'])
+        error = exc_info[1]
+        qdb.logger.LogEntry.create(
+            'Runtime',
+            'ERROR:\n%s\nTRACE:\n%s\nHTTP INFO:\n%s\n' %
+            (error, trace_info, request_info))
+
+    # Allow call to oauth_error function as part of the class
+    oauth_error = _oauth_error
+
+    def head(self):
+        """Adds proper response for head requests"""
+        self.finish()
+
+
+class TokenAuthHandler(OauthBaseHandler):
+    def generate_access_token(self):
+        """Creates the random alphanumeric token
+
+        Returns
+        -------
+        str
+            55 character random alphanumeric string
+
+        Notes
+        -----
+        55 was chosen as a cryptographically secure limit that needs to be
```

Comment:
Ok, can you actually add the comment that the exact length of the token can change, but it should never be longer than 255? That is a technology limit and it will be good to have it exposed here in the documentation.

I don't see any check in here but, what happens if the token that is generated already exists in the system? The RFC does not say anything specific, but I think we should double check that each new access_token is unique, or am I missing something?